rollingpin/elasticsearch.py

```diff
@@ -47,20 +47,62 @@ def __init__(self, config, components, hosts, command_line, word):
         self.endpoint = "https://%s/%s/%s" % (base_url, index, index_type)
         self.components = components
 
-    def index_doc(self, doc):
+    def index_doc(self, doc, cb=None):
         """ Index a document in Elasticsearch
         :param doc: dictionary with data to index in ES
+        :param cb: an optional callback that will be added to the
+            ``Agent.request`` deferred.  Will be called with an instance of
+            ``twisted.web.client.Response`` as the first argument.
         :return: Deferred
         """
         with swallow_exceptions("elasticsearch", self.logger):
             agent = Agent(reactor)
             body = JSONBodyProducer(json.dumps(doc))
-            return agent.request(
+            d = agent.request(
                 'POST',
                 self.endpoint,
                 Headers({'User-Agent': ['rollingpin']}),
                 body,
             )
+            if cb:
+                d.addCallback(cb)
+            return d
+
+    def update_doc(self, id_, updated_fields):
+        """ Update a document in Elasticsearch
+        :param updated_fields: dictionary values to update
+        :return Deferred
+        """
+        # The twisted web client chokes on unicode URLs, so we'll coerce this
+        # back to a string
+        update_endpoint = str("%s/%s/_update" % (self.endpoint, id_))
+
+        # Unlike indexing, for updates we have to put our updated fields inside
+        # a "doc" key.
+        #
+        # https://www.elastic.co/guide/en/elasticsearch/guide/current/partial-updates.html
+        doc = {'doc': updated_fields}
+
+        with swallow_exceptions("elasticsearch", self.logger):
+            agent = Agent(reactor)
+            body = JSONBodyProducer(json.dumps(doc))
+            return agent.request(
+                'POST',
+                update_endpoint,
+                Headers({'User-Agent': ['rollingpin']}),
+                body,
+            )
```

Comment:
My usual philosophy on it is 1) first time, write it, 2) second time, copy it, 3) third time, refactor.  This has enough differences between the endpoint, document structure, and usage of callbacks that I'm not sure the part that's being reused is really anything more than a single function call.